backend/src/main/java/com/google/sps/servlets/AddParticipantServlet.java

```diff
@@ -47,51 +49,68 @@
+  private static final String REQUEST_ROLE = "role";
+  private static final String REQUEST_PRODUCT_AREA = "productArea";
+  private static final String REQUEST_SAVE_PREFERENCE = "savePreference";
+  private static final String REQUEST_MATCH_PREFERENCE = "matchPreference";
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Request parameter values
-    // TODO: Check input values are filled (before allowing to submit)
-    UserService userService = UserServiceFactory.getUserService();
-    String email = userService.getCurrentUser().getEmail();
-    if (email == null) {
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid email.");
+    JSONObject obj = retrieveRequestBody(request);
+    if (obj == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Could not read request body");
       return;
     }
-    String username = email.split("@")[0];
+    JSONObject formDetails = obj.getJSONObject(REQUEST_FORM_DETAILS);
+
+    // Retrieve the timeAvailableUntil input and convert to a UTC ZonedDateTime
+    long timeAvailableUntil = formDetails.getLong(REQUEST_TIME_AVAILABLE_UNTIL);
+    ZoneId zoneId = ZoneId.of("UTC");
+    ZonedDateTime startTimeAvailable = ZonedDateTime.now(Clock.systemUTC());
+    ZonedDateTime endTimeAvailable =
+        ZonedDateTime.ofInstant(Instant.ofEpochMilli(timeAvailableUntil), zoneId);
 
-    String timezone = request.getParameter("timezone");
-    ZoneId zoneId = ZoneId.of(timezone); // TODO: convert input timezone to valid ZoneId
-    ZonedDateTime startTimeAvailable =
-        ZonedDateTime.now(zoneId); // TODO: set to future time if not available now
-    Instant endTimeAvailableInstant =
-        Instant.ofEpochMilli(
-            Long.parseLong(
-                request.getParameter("endTimeAvailable"))); // TODO: figure out input format
-    ZonedDateTime endTimeAvailable = endTimeAvailableInstant.atZone(zoneId);
-
-    int duration = convertToPositiveInt(request.getParameter("duration"));
+    int duration = formDetails.getInt(REQUEST_DURATION);
     if (duration <= 0) {
       response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid duration.");
       return;
     }
 
+    String role = formDetails.getString(REQUEST_ROLE);
+    String productArea = formDetails.getString(REQUEST_PRODUCT_AREA);
+    boolean savePreference = formDetails.getBoolean(REQUEST_SAVE_PREFERENCE);
+    String matchPreference = formDetails.getString(REQUEST_MATCH_PREFERENCE);
+
     long timestamp = System.currentTimeMillis();
 
+    // Retrieve user email address via Users API and parse for ldap
+    UserService userService = UserServiceFactory.getUserService();
+    String email = userService.getCurrentUser().getEmail();
+    if (email == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid email.");
+      return;
+    }
+    String username = email.split("@")[0];
```

Comment:
I think there's still an opportunity to create a function for this part, as something like `getInfoFromJSONObject`, so we can further reduce the body of `doPost`. I am OK with refactoring in another PR, but I'm worried this function getting much larger (and harder to understand) when we add more and more functionalities.

Reply:
Modularize this functionality into its own method, thanks!